### PR TITLE
[27] - Create messages page

### PR DESCRIPTION
**Pull request by otash8110**
### Where

[NL-27 Create messages page](https://jjbcapital.atlassian.net/browse/NL-27)

### What

Implement chat functionality and page on frontend

### How

Using updates provided on backend side, create frontend layout for chats and messaging. this branch populates 1 conversatio between seller@localhost.com and buyer@localhost.com for testing purposes. All real conversations will be created by background job workers when lot statuses are changed to sold

### Affected Area

Auctionify.API.ClientApp
Auctionify.Application

### PR Checklist

_Build_

- [x] Build Succeeded

_Unit Tests_

- [x] 102 passed, 0 Ignored, 0 Failed

### CODE REVIEW

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                                subscriber.unsubscribe();
+                            }
+                        );
+                    }
+                },
+            });
+    }
+
+    scrollBehaviour(el: MessageObserverDirective | null) {
+        console.log(el, this.firstUnreadId);
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts:129]: Do we need to log this info into console?

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                (group) => group.date === dateKey
+            );
+
+            if (existingGroup) {
+                if (
+                    this.isFirstUnread &&
+                    !message.isRead &&
+                    message.senderId != this.currentUserId
+                ) {
+                    console.log('UNREAD');
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts:184]: do we need it here too?

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+    generateMessageClass(message: ChatMessage) {
+        if (this.currentUserId == message.senderId) {
+            return 'self';
+        } else return 'other';
+    }
+
+    intersect(event: IntersectResult) {
+        if (event.isIntersecting && !event.isRead && !event.isOwner) {
+            console.log(event.isIntersecting, event.isRead);
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts:226]: what about here?

**Comment by khasanrashidov** [comment]
When there are no conversations (which means there are no chats), I think the following should be shown

image

What we have now (the case we don't have conversations):

image

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/chat-page/chat-page.component.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        this.chosenConversation = this.chatConversations.find(
+            (x) => x.id == id
+        )!;
+    }
+
+    getMainChatContainerWidthClass() {
+        if (this.isUserBuyer) {
+            return 'chat-width-buyer';
+        } else {
+            return 'chat-width-seller';
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/chat-page/chat-page.component.ts:91]: I would assign them into constants and return them

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                    subscriber.unsubscribe();
+                });
+            },
+        });
+    }
+
+    generateMessageClass(message: ChatMessage) {
+        if (this.currentUserId == message.senderId) {
+            return 'self';
+        } else return 'other';
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts:221]: is it better to assign them into constants?

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        ) {
+            this.conversation = changes['conversation'].currentValue;
+            this.getAllConversationMessages();
+        }
+    }
+
+    ngAfterViewInit(): void {
+        this.mainConversationWindow.nativeElement.onscroll = (el) => {
+            let percent = this.calculateCurrentScrollPercent();
+            if (percent < 97.6) this.isScrolledUp = true;
```
Comment (khasanrashidov) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/conversation-window/conversation-window.component.ts:86]: As I understand, 97.6 is the threshold for the scroll to be considered at the bottom, right?

**Review by khasanrashidov — APPROVED** [review]

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/chats/chats.component.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import {
+    Component,
+    EventEmitter,
+    Input,
+    OnChanges,
+    OnInit,
```
Comment (Muhammad-gf) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/chats/chats.component.ts:6]: You didn't use ngOnInit. You can remove it.

File: src/Auctionify.API/ClientApp/src/app/components/general/chat/chat-page/chat-page.component.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import {
+    Component,
+    ElementRef,
```
Comment (Muhammad-gf) [inline src/Auctionify.API/ClientApp/src/app/components/general/chat/chat-page/chat-page.component.ts:3]: Remove 'ElementRef, ViewChild, ChatsComponent'.

**Review by Muhammad-gf — CHANGES_REQUESTED** [review]

**Review by khasanrashidov — DISMISSED** [review]

**Review by Muhammad-gf — APPROVED** [review]

**Review by SaidganiUsm — APPROVED** [review]

**Review by khasanrashidov — APPROVED** [review]

**Pull request merged**